build-std: always

davidtwco opened the pull request:
*[View all comments](https://triagebot.infra.rust-lang.org/gh-comments/rust-lang/rfcs/pull/3874)*

Add a new Cargo configuration option, `build-std = "always|never"`, which will unconditionally rebuild standard library dependencies. The set of standard library dependencies can optionally be customised with a new `build-std-crates` option. It also describes how Cargo (or external tools) should build the standard library crates on stable (i.e., which flags to pass and features to enable).

This proposal limits the ways the built standard library can be customised (such as by settings in the profile) and intends that the build standard library matches the prebuilt one (if available) as closely as possible.

**This RFC is is part of the [build-std project goal](https://rust-lang.github.io/rust-project-goals/2025h2/build-std.html) and a series of build-std RFCs:**

1. build-std context (rust-lang/rfcs#3873)
2. `build-std="always"` (this RFC)
3. Explicit standard library dependencies (rust-lang/rfcs#3875)
4. `build-std="compatible"` (RFC not opened yet)
5. `build-std="match-profile"` (RFC not opened yet)

There is also [a Zulip channel where you can ask questions about any of the build-std RFCs](https://rust-lang.zulipchat.com/#narrow/channel/516120-project-goals.2F2025h1.2Fbuild-std). This series of RFCs was drafted over many months with the help of stakeholders from many Rust project teams, we thank them for their help!

[FCP](https://github.com/rust-lang/rfcs/pull/3874#issuecomment-3715844925)

[Rendered](https://github.com/rust-lang/rfcs/blob/master/text/3874-build-std-always.md)

davidtwco committed:
build-std: always

davidtwco committed:
rename to rfc 3874

davidtwco committed:
update forward links

clarfonthey commented on text/3874-build-std-always.md:
```diff
@@ -0,0 +1,1538 @@
+
+  - `panic_abort` is a non-optional dependency of `std` so it will always be
+    built
+
```
Worth commenting on `-Cpanic=immediate-abort` which was recently added: rust-lang/rust#146974

clarfonthey commented on text/3874-build-std-always.md:
```diff
@@ -0,0 +1,1538 @@
+## Panic strategies
+[panic-strategies]: #panic-strategies
+
+Panic strategies are unlike other profile settings insofar as they influence
```
Putting this here since there's not a great place to put it: while it's okay to make a future addition, I think it's good to mention the `optimize_for_size` feature as one of the other main things that people care about when building `std`. There are potential flags we could check for whether to enable this feature, but it's probably going to want its own dedicated flag associated with `build-std` and thus we shouldn't infer it here.

For example, I can imagine `-O2` plus `optimize_for_size` being useful because you're willing to permit, for example, loop unrolling or autovectorization, but not large tables being included in the binary. So, simply checking for `-Os` or `-Oz` is not adequate.

epage commented on text/3874-build-std-always.md:
```diff
@@ -0,0 +1,1538 @@
+[unresolved-config-name]: #what-should-the-build-std-configuration-in-cargoconfig-be-named
+
+What should this configuration option be named? `build-std`?
+`rebuild-standard-library`?
```
I thought we had notes on considerations for this?

For instance, `build-std = "always"` could be misinterpreted as "don't apply rebuild rules"

epage commented on text/3874-build-std-always.md:
````diff
@@ -0,0 +1,1538 @@
+
+```toml
+[build]
+build-std-crates = "std"
````
A related thought I had when seeing this field by itself is it could be interpreted to mean `build-std = "always"` on its own.  Having this example be more complete would help but it could still be confusing when layering different configs.

Unsure if its fine on its own, needs a different structure, a warning or note, or something else.

epage commented on text/3874-build-std-always.md:
```diff
@@ -0,0 +1,1538 @@
+
+- [*Allow building `profiler-builtins`*][future-profiler-builtins]
+
+## Caching
```
`build-dir` does not change the situation regarding caching, only where they are located.  A user can force `build-dir` to be shared like they can for `target-dir` today but it is not expected to be a default or generally encouraged pattern.

tgross35 commented on text/3874-build-std-always.md:
```diff
@@ -0,0 +1,1538 @@
+apply to the opaque dependency. However, if a target modifier were set using
+`RUSTFLAGS` and build-std ignored the variable, then rustc would fail to build
+the user's project due to incompatible target modifiers. This would necessitate
+that every stable target modifier be exposed via Cargo to be usable in practice.
```
I have run into a friction point with the current build-std in that setting `RUSTFLAGS=-Dwarnings` sometimes causes the std build to fail on T3 targets. It may be worth noting somewhere that std's lints will be suppressed and not be affected by RUSTFLAGS, similar to other dependencies.

davidtwco replied:
I've mentioned that we'll cap-lints for standard library crates in 4db3d98.

davidtwco committed:
add oxford commas

davidtwco committed:
cap lints in std crates

davidtwco committed:
move links to relevant sections

davidtwco committed:
features are in the next rfc

davidtwco committed:
move no-std future possibility to next rfc

davidtwco committed:
-Cpanic=immediate-abort

davidtwco committed:
add object form of build-std

davidtwco commented on text/3874-build-std-always.md:
```diff
@@ -0,0 +1,1538 @@
+`metadata.std` field. `standard_library_support` has two fields:
+
+- `supported`, which can be set to either "core", "core and alloc", or
+  "core, alloc and std"
```
Fixed in 3f1b89b

epage commented on text/3874-build-std-always.md:
```diff
@@ -0,0 +1,1543 @@
+
+> [!NOTE]
+>
+> Inspired by the concept of [opaque dependencies][Opaque dependencies], the
```
@nicoburns brought up on reddit the idea of `cargo clean` not affecting opaque dependencies, requiring some other step to clean them (maybe a `cargo clean -p std`).  Alternatively, we may get this "for free" with rust-lang/cargo#5931.

I lean towards keeping `cargo clean` the same for now but might be worth calling this question out somewhere.

ehuss committed:
Increase heading levels

As part of https://github.com/rust-lang/rfcs/pull/3883, we have switched
the template to stop using level-1 headings which isn't the way mdbook
is intended to be used.

bjorn3 commented on text/3874-build-std-always.md:
```diff
@@ -0,0 +1,1543 @@
+invoked ([?][rationale-implied-bootstrap]). Cargo will not need to use
+`RUSTC_BOOTSTRAP` when compiling the standard library with a stable toolchain.
+The standard library's dependencies will not be permitted to use build probes to
+detect whether a nightly version is being used.
```
I went through all build scripts in dependencies we have. The libc crate detects both the rustc version and if it is a nightly version. It seems like it doesn't actually use the latter for anything though. The object crate also detects rustc version. Nothing else seems to be detecting the rustc version that I could see.

davidtwco replied:
I've added a note that `libc` and `object` are the only affected crates at the moment in a0977f78

bjorn3 commented on text/3874-build-std-always.md:
```diff
@@ -0,0 +1,1543 @@
+Ideally, a crate should be able to provide alternate memory symbols and disable
+`compiler_builtins`' symbols for the entire crate graph by enabling a feature
+(e.g. `std`/`libc` could do this) - this is what an `external-mem` feature
+enables.
```
Until external-mem automatically gets enabled for all targets with a libc implementation, this inversion of the mem feature would cause a silent performance hit for those targets, while previously you would get a loud failure, which IMO is preferable.

bjorn3 commented on text/3874-build-std-always.md:
```diff
@@ -0,0 +1,1543 @@
+change (e.g. switching between stable and nightly and back). The `dylib` is only
+linked against when `-Cprefer-dynamic` is used. build-std will initially be
+conservative and not include the `dylib` and `-Cprefer-dynamic` would fail
+compilation.
```
I'm pretty sure `-Cprefer-dynamic` would fallback to static linking rather than failing compilation.

davidtwco committed:
update opaque deps link

davidtwco committed:
update links to merged rfc

ehuss commented on text/3874-build-std-always.md:
```diff
@@ -0,0 +1,1543 @@
+the standard library, the generated documentation for the user's crates will
+link to the locally generated documentation for the `core`, `alloc` and `std`
+crates, rather than the upstream hosted generation as is typical for non-locally
+built standard libraries.
```
This seems like an OK compromise to start, but I'm not entirely comfortable with it. The standard library documentation takes half a gigabyte of disk space, and can take quite a while to generate.

In the future we may want to improve this situation somehow. Cargo could use the same logic that rustdoc uses to find the correct channel (I think it's around [here](https://github.com/rust-lang/rust/blob/4fa80a5e733e2202d7ca4c203c2fdfda41cfe7dc/src/librustdoc/clean/utils.rs#L580-L586)). Cargo already generates channel-specific links to doc.rust-lang.org. Or maybe there is some extension to rustdoc that we could use.

For context, the issue tracking this problem is https://github.com/rust-lang/wg-cargo-std-aware/issues/88.

Cargo also has an unstable [rustdoc-map](https://doc.rust-lang.org/nightly/cargo/reference/unstable.html#rustdoc-map) feature for controlling where the std docs link to. Perhaps we could lean in on that. Unfortunately it's stuck on a bunch of issues, and one important one is [the precedence of `--extern-html-root-url` and `--extern-html-root-takes-precedence`](https://github.com/rust-lang/rust/pull/82776).

davidtwco replied:
I very much intend what's written to be the conservative "this will definitely work" approach and would be happy to see us try and do something better here.

ehuss commented on text/3874-build-std-always.md:
```diff
@@ -0,0 +1,1543 @@
+Vendoring the standard library is possible since it currently has its own
+workspace, allowing the dependencies of just the standard library crates (and
+not the compiler or associated tools in `rust-lang/rust`) to be easily packaged.
+Doing so has multiple advantages..
```
Yea, I'm not sure how we're going to implement this. I'm a little concerned it will be tricky, and is good for us to keep in mind that this will be a challenging change.

davidtwco committed:
unresolved q for `cargo clean`

davidtwco committed:
mention current deps w/ build probes

davidtwco committed:
prefer-dynamic static fallback

davidtwco committed:
json output for target-standard-library-support

bjorn3 commented on text/3874-build-std-always.md:
```diff
@@ -0,0 +1,1543 @@
+[preventing-implicit-sysroot-dependencies]: #preventing-implicit-sysroot-dependencies
+
+Cargo will pass a new flag to rustc which will prevent rustc from loading
+top-level dependencies from the sysroot ([?][rationale-root-sysroot-deps]).
```
build-std would recompile the entire standard library, right? So there are no dependencies that need to be loaded from the sysroot.

JohnDowson replied:
https://github.com/rust-lang/rfcs/pull/3874#discussion_r2534912578

`up-to`name making sense suggests to me that it is possible to rebuild `std` without rebuilding `core`+`alloc`?

bjorn3 replied:
The way I interpret is that `up-to = "std"` would rebuild `core`, `alloc` and `std`, but not any later crates like `proc_macro` or `test`.

epage commented on text/3874-build-std-always.md:
````diff
@@ -0,0 +1,1567 @@
+```toml
+[build]
+build-std = { when = "always" }
+```
````
Something that came up with `config-include` is deciding whether the primary form of a table should be a standard table or inline table and that can help us decide whether short-hands are appropriate or not.

Is there a reason you are using an inline table rather than a standard table?

davidtwco replied:
Not especially, just for conciseness in the RFC text. My gut feeling is that `build-std.when` would be most common because it'll eventually be the only option that users need to set.

epage commented on text/3874-build-std-always.md:
````diff
@@ -0,0 +1,1543 @@
+
+```toml
+[build]
+build-std = { when = "always", crates = "std" }
````
I would like to see the confusing name of a plural for a single value at least brought up in the "what to name" so we don't lose track of what the concern is.

joshtriplett replied:
Another possible alternative name: `which`, to go with `when`.

`up-to` feels really quirky.

epage commented on text/3874-build-std-always.md:
```diff
@@ -0,0 +1,1567 @@
+dependencies will be used if `build-std.crates` is not explicitly set.
+Otherwise, `build-std.crates` will default to the crate intended to be supported
+by the target (see later
+[*Standard library crate stability*][standard-library-crate-stability] section).
```
This reads in a confusing way where the higher precedent item is given in an "otherwise it will default" branch of the logic.

davidtwco committed:
elaborate on `build-std.crates` unresolved question

davidtwco committed:
move `?` to more accurate location

davidtwco committed:
clarified behaviour for unset `build-std.crates`

epage commented:
@rfcbot fcp concern cargo-config-shorthand

See https://github.com/rust-lang/rfcs/pull/3874#discussion_r2670304203

davidtwco committed:
remove shorthand

davidtwco committed:
mention `nounused`

davidtwco commented on text/3874-build-std-always.md:
```diff
@@ -0,0 +1,1575 @@
+> directory of the crate or workspace like any other dependency.
+>
+> Standard library crates are provided to the compiler using the `--extern` flag
+> with the `noprelude` modifier ([?][rationale-noprelude-with-extern]).
```
Mentioned this in 39738034

epage commented:
@rfcbot fcp concern rustflags-profile-inconsistency

See https://github.com/rust-lang/rfcs/pull/3874/changes#r2879178332

davidtwco committed:
link to future proposals

davidtwco committed:
elaborate on motivation for future proposals

davidtwco committed:
build-std is a drop-in replacement

davidtwco committed:
elaborate on rustflags/profile inconsistency

davidtwco committed:
elaborate further on build-std drop-in replacement

joshtriplett commented on text/3874-build-std-always.md:
```diff
@@ -0,0 +1,1631 @@
+
+What should this configuration option be named? In particular, `crates` being
+plural is unintuitive: while it can enable building multiple crates, it is set
+with the name of only a single crate. `up-to` might be better.
```
```suggestion
with the name of only a single crate. `up-to` or `which` might be better.
```

davidtwco replied:
Added in 801e1e35

joshtriplett commented on text/3874-build-std-always.md:
```diff
@@ -0,0 +1,1631 @@
+> - Lints in standard library crates will be built using `--cap-lints allow`
+>   matching other upstream dependencies.
+>
+> Cargo will resolves the dependencies of opaque dependencies, such as the
```
```suggestion
> Cargo will resolve the dependencies of opaque dependencies, such as the
```

davidtwco replied:
Fixed in 801e1e35

joshtriplett commented on text/3874-build-std-always.md:
```diff
@@ -0,0 +1,1631 @@
+> Standard library crates are provided to the compiler using the `--extern` flag
+> with the `noprelude` modifier ([?][rationale-noprelude-with-extern]). Standard
+> library crates are also provided with the `nounused` modifier to avoid being
+> considered an unused crate dependency.
```
Please consider adding a rationale for `nounused`.

davidtwco replied:
Added in 9171316d

joshtriplett commented on text/3874-build-std-always.md:
```diff
@@ -0,0 +1,1631 @@
+`metadata.std` field. `standard_library_support` has two fields:
+
+- `supported`, which can be set to either "core", "core and alloc", or
+  "core, alloc, and std"
```
```suggestion
- `supported`, which can be set to either "core", "core, alloc", or
  "core, alloc, std"
```

davidtwco replied:
Fixed in 801e1e35

joshtriplett commented on text/3874-build-std-always.md:
```diff
@@ -0,0 +1,1631 @@
+- `default`, which can be set to either "core", "core and alloc", or
+  "core, alloc, and std"
+  - `default` cannot be set to a value which is "less than" that of `supported`
+    (i.e. "core and alloc" when `supported` was only set to "core")
```
```suggestion
- `default`, which can be set to either "core", "core,  alloc", or
  "core, alloc, std"
  - `default` cannot be set to a value which is "less than" that of `supported`
    (i.e. "core, alloc" when `supported` was only set to "core")
```

davidtwco replied:
Fixed in 801e1e35

joshtriplett commented on text/3874-build-std-always.md:
```diff
@@ -0,0 +1,1631 @@
+
+Cargo's `build-std.crates` field will default to the value of the
+`standard_library_support.default` field (`std` for "core, alloc, and std",
+`alloc` for "core and alloc", and `core` for "core"). This does not prevent
```
```suggestion
`standard_library_support.default` field (`std` for "core, alloc, std",
`alloc` for "core, alloc", and `core` for "core"). This does not prevent
```

davidtwco replied:
Fixed in 801e1e35

joshtriplett commented on text/3874-build-std-always.md:
```diff
@@ -0,0 +1,1631 @@
+[unresolved-config-name]: #what-should-the-build-stdwhen-configuration-in-cargoconfig-be-named
+
+What should this configuration option be named? `build-std`?
+`rebuild-standard-library`?
```
I think `when` is a fine name.

joshtriplett commented:
@rfcbot reviewed

Thank you for a very clear and detailed writeup!

davidtwco committed:
minor grammatical/typo fixes

davidtwco committed:
add rationale for nounused

epage commented on text/3874-build-std-always.md:
```diff
@@ -0,0 +1,1580 @@
+apply to the opaque dependency. However, if a target modifier were set using
+`RUSTFLAGS` and build-std ignored the variable, then rustc would fail to build
+the user's project due to incompatible target modifiers. This would necessitate
+that every stable target modifier be exposed via Cargo to be usable in practice.
```
When talking about target modifiers, it is an exception and one I would not be thrilled with stabilizing.  If we're taking the stance though that we're blindly applying RUSTFLAGS to opaque dependencies and the way to avoid that is to find ways to model RUSTFLAGS in other parts of Cargo, like profiles, then that is hopefully good enough and not something we'll come to regret.

epage commented:
@rfcbot fcp resolve rustflags-profile-inconsistency

epage commented on text/3874-build-std-always.md:
```diff
@@ -0,0 +1,1580 @@
+there is less chance of the user experiencing bugs or unexpected behaviour from
+the well-tested and supported pre-built std. Later proposals will extend the
+`build-std` option with customised standard library builds that use the user's
+profile.
```
I wish I had written more about this, how much is build-std specific vs opaque dependencies.  In general, I want us to be making this just a built-in opaque dependency and there to be principled ways of applying these ideas.

epage commented:
@rfcbot fcp resolved user-profile

weihanglo commented on text/3874-build-std-always.md:
```diff
@@ -0,0 +1,1644 @@
+
+*See the following sections for relevant unresolved questions:*
+
+- [*What should the `build-std` configuration in `.cargo/config` be named?*][unresolved-config-name]
```
nit: `.cargo/config` (without `.toml` extension`) is not preferred

```suggestion
- [*What should the `build-std` configuration in `.cargo/config.toml` be named?*][unresolved-config-name]
```

davidtwco replied:
Fixed in 237f16e3

weihanglo commented on text/3874-build-std-always.md:
```diff
@@ -0,0 +1,1644 @@
+unconditionally recompiled ([?][rationale-unconditional]) in the release profile
+([?][rationale-release-profile]) defined in its workspace as part of every clean
+build. This is primarily useful for users of tier three targets. As with other
+dependencies, the standard library's build will respect the `RUSTFLAGS`
```
When we say `RUSTFLAGS`, I guess we refer to [all different stable ways to specify rustflags in Cargo](https://doc.rust-lang.org/nightly/cargo/reference/config.html#buildrustflags), their precedence, right?

> * env `CARGO_ENCODED_RUSTFLAGS`
> * env `RUSTFLAGS`
> * config `target.<triple>.rustflags` (including its env equivalent)
> * config `target.<cfg>.rustflags` (including its env equivalent)
> * config `build.rustflags (including its env equivalent)

weihanglo replied:
(As we don't respect profile right now, `profile.rustflags` unstable features won't be applied. Also we can leave interactions with other nightly features `target-applies-to-host` for future)

davidtwco replied:
I think it makes sense that this would apply to all of the ways that `RUSTFLAGS` gets set, and that `profile.rustflags` wouldn't apply (from the user's profile, but it probably should from the profile that the build uses - i.e. the standard library's release profile). Fixed in b87180cb.

weihanglo replied:
Thanks for the update!

weihanglo commented on text/3874-build-std-always.md:
```diff
@@ -0,0 +1,1644 @@
+[caching]: #caching
+
+Standard library artifacts built by build-std will be reused equivalently to
+today's crates/dependencies that are built within a shared target directory. By
```
This might be an implementation detail open to change after this RFC merge, though I wonder whether `libstd.rlib` is a final artifact or intermediate artifacts. As noted earlier in some review comment, Cargo now splits out build-directory (intermediate artifacts, not accessible by end user and its layout is unstable) from target directory.

I guess we consider it intermediate artifact, right? So this RFC will not allow user to abuse `build-std` as a way to build and distribute libstd artifacts.

bjorn3 replied:
If you dynamically link libstd it did have to be uplifted, otherwise I agree it should be intermediate artifacts.

davidtwco replied:
Agreed with both - noted in c55d12f.

davidtwco committed:
s#.cargo/config#.cargo/config.toml#

davidtwco committed:
clarify all methods of setting rustflags

davidtwco committed:
std rlibs are intermediate artifacts

ehuss committed:
Add tracking issue

ehuss commented:
Thanks everyone! The tracking issue is https://github.com/rust-lang/rust/issues/155363.
